### DISCUSSION TITLE
Trouble getting expected results with Shelly Plus 1 with Addon for contact sensor

### DISCUSSION THREAD

**Discussion by BadgerBadgerAndFox** [arachnetech/homebridge-mqttthing#611] [Q&A]
I have just started using this plugin in Homebridge and am not getting the expected result for contact sensor configuration.

I have deployed a couple of Addon modules to existing Shelly Plus 1 modules with DHT22 sensors and a reed switch.
The temperature/humidity sensors are working, although i have noted a small issue particularly with humidity in that the plugin reports no value until a message is sent to the MQTT broker which i think only occurs on a state change. Not 100% sure on this as i am new to MQTT.

This results in the humidity sensor reporting no value for one of the sensors that is in a very stable location. Is it possible to present a historical value on the startup?

My primary query however is with the contact sensor. The logs indicate that the topic is correctly returning states of true/false when the contact senso transitions between states, however the sensor state in homebridge does not change state and with history enabled the eve app also indicates no state changes have occurred. Am i missing something in the config?

        {
            "type": "contactSensor",
            "name": "Entry Door",
            "url": "mqtt://MqttServer:1883",
            "logMqtt": true,
            "topics": {
                "getContactSensorState": {
                    "topic": "shellyplus1-MacAddress/status/input:100$.state"
                }
            },
            "history": true,
            "accessory": "mqttthing"
        },

Below is the message for the topic as observed for the sensor being closed in MQTT Explorer

`{
  "id": 100,
  "state": true
}`

Below is the message for the topic as observed for the sensor being opened in MQTT Explorer

`{
  "id": 100,
  "state": false
}`

Homebridge logs correctly report the state changes as below

jsonpath $.state decoded message to [true]
jsonpath $.state decoded message to [false]

Any ideas where i am going wrong?

**Comment by BadgerBadgerAndFox** [comment]
Would the retain feature of the json codec assist with setting the initial value of the sensors by allowing the last state to be reported rather than waiting for an updated state change?

`"jsonCodec": {
    "retain": true
}`

**Comment by HolgerAusB** [comment]
I am also new with zigbee2mqttt, but as far as I found out, the message payload is in JSON, so you need to parse this. You need to edit the sensor  config by hand, e.g.:

I needet to negate the JSON.parse by the **`!`** because HB was showing window open when it was closed and vice versa.

```
{
    "type": "contactSensor",
    "name": "MySensorName",
    "url": "mqtt://xxx",
    "logMqtt": true,
    "topics": {
        "getContactSensorState": {
            "topic": "zigbee2mqtt/MySensor",
            "apply": "return !JSON.parse(message).contact;"
        },
        "getBatteryLevel": {
            "topic": "zigbee2mqtt/MySensor",
            "apply": "return JSON.parse(message).battery;"
        }
    },
    "accessory": "mqttthing"
}
```

**Reply by BadgerBadgerAndFox** [reply]
You can skip the parse via apply if you leverage Jsonpath as per the config guide.
I’m doing this successfully for humidity and temperature sensors.

as per my homebridge log I am getting the expected value of true or false, issue is that it does not trigger a state change in homebridge

**Reply by HolgerAusB** [reply]
Interesting! But I can't find that guide. Please link to that.

**Reply by HolgerAusB** [reply]
But instead I found out how to change the output in zigbee2mqtt.io if you are using this, you can add following lines to your configuration.yaml:

```
advanced:
  output: attribute
```
Puts every parameter to seperate subtopics: zigbee2mqtt/MySensor/contact with a simple true/false as payload

You also can have both formats with 

```
advanced:
  output: attribute_and_json
```

**Reply by BadgerBadgerAndFox** [reply]
This is the state change logged in Homebridge for the contact sensor, the values seem to be correct and i have even added the expected optional values as per the doco (which should not be required as they are the defaults)

[23/02/2023, 08:45:05] [Entry Door] Received MQTT: shellyplus1-xx/status/input:100 = {"id":100,"state":false}
[23/02/2023, 08:45:05] [Entry Door] jsonpath $.state decoded message to [false]
[23/02/2023, 08:45:15] [Entry Door] Received MQTT: shellyplus1-xx/status/input:100 = {"id":100,"state":true}
[23/02/2023, 08:45:15] [Entry Door] jsonpath $.state decoded message to [true]

This is the config as currently applied:

        {
            "type": "contactSensor",
            "name": "Entry Door",
            "url": "mqtt://server:1883",
            "logMqtt": true,
            "topics": {
                "getContactSensorState": {
                    "topic": "shellyplus1-xx/status/input:100$.state"
                }
            },
            "integerValue": false,
            "onValue": "true",
            "offValue": "false",
            "onlineValue": "true",
            "offlineValue": "false",
            "history": true,
            "accessory": "mqttthing"
        }

I think the issue may be that its not accepting the value as true/false as its not boolean... trying to figure out how to parse the value to boolean. I will try flipping configs between having just intergerValue:false and setting onvalue/offvalue as the doco indicates it should be one or the other not both. https://github.com/arachnetech/homebridge-mqttthing/blob/master/docs/Configuration.md#boolean-value-settings

The jsonpath doco is here: https://github.com/arachnetech/homebridge-mqttthing/blob/master/docs/Configuration.md#jsonpath

**Reply by BadgerBadgerAndFox** [reply]
ok... that fixed it :)

    {
        "type": "contactSensor",
        "name": "Entry Door",
        "url": "mqtt://server:1883",
        "logMqtt": true,
        "topics": {
            "getContactSensorState": {
                "topic": "shellyplus1-xx/status/input:100$.state"
            }
        },
        "integerValue": false,
        "onlineValue": "true",
        "offlineValue": "false",
        "history": true,
        "accessory": "mqttthing"
    }


Removed the following lines...

        "onValue": "true",
        "offValue": "false",

Now i just need to work out the scripts to run locally on the shelly's to publish temp/humidity topics every 30 seconds to deal with infrequent updates in stable environmnets... I was hoping to enable retain on the shelly's MQTT config, but they removed that option in the gen2 devices, it was possible in gen1 devices but they recommended against it.

**Reply by ukiews** [reply]
@BadgerBadgerAndFox Have you been able to figure out the scripting to report regular temp and humidity? If so, could you please share? I am using the DHT22 with the 1PMG3.
